### ISSUE TITLE
bug: tasks are run before and after generating or updating a project when prompt is configured

### ISSUE THREAD

**Issue by esciara** [copier-org/copier#499]
**Describe the problem**

Tasks are run before and after generating or updating a project, where as they are supposed to run only once after generating or updating a project.

The [doc says](https://copier.readthedocs.io/en/stable/configuring/#tasks) that `Commands to execute after generating or updating a project from your template.` And having the task run twice with it is also templated can lead to unattended consequences: in one case, depending on the prompt, a directory would be deleted or not. In particular here, when the task runs a python script file for instance, and this file has been changed, it is the old version of the script that is run before the generation and the new one that is run after the generation.

**Template**

Here is the [copier-troubleshoot-task.zip](https://github.com/copier-org/copier/files/7672216/copier-troubleshoot-task.zip)  file containing a `copier` template to reproduce the problem.

**To Reproduce**

1. Install the stated version of `copier` as described (see **Environment** section)
2. Unzip the sample template in a directory (here, we will use `copier-troubleshoot-task/`). You will obtain the following tree:
    ```
    copier-troubleshoot-task
    ├── copier.yaml
    └── project
        ├── [[\ _copier_conf.answers_file\ ]].tmpl
        └── copier_post_generation_task.py
    ```
3. Follow the command lines below to generate a project from the template and setup a git repo for the generated content:
    <details>

    ```bash
    $ cd copier-troubleshoot-task/
    $ git init
    $ git add .
    $ git commit -m'first commit'
    [main (root-commit) c9f73e2] first commit
    3 files changed, 6 insertions(+)
    create mode 100644 copier.yaml
    create mode 100644 project/[[ _copier_conf.answers_file ]].tmpl
    create mode 100644 project/copier_post_generation_task.py
    $ copier . ../copier-troubleshoot-task-gen
    No git tags found in template; using HEAD as ref

        create  copier_post_generation_task.py
        create  .copier-answers.yml

    > Running task 1 of 1: python3 copier_post_generation_task.py
    You ran 'copier_post_generation_task.py'!

    $ pushd ../copier-troubleshoot-task-gen
    $ git init
    $ git add .
    $ git commit -m'first template generation'
    [main (root-commit) b50a583] first template generation
    2 files changed, 3 insertions(+)
    create mode 100644 .copier-answers.yml
    create mode 100644 copier_post_generation_task.py
    $ popd
    ```
    </details>
4. Change the `copier_post_generation_task.py`:
    ```python3
    import os
    
    if __name__ == "__main__":
    print(f"You really ran '{os.path.basename(__file__)}'!")
    ```
5. Re-generate the template, and note that the task is only run after the files are generated:
    <details>

    ```bash
    # In the `copier-troubleshoot-task/` directory
    $ git add .
    $ git commit -m'update task script'
    [main (root-commit) c9f73e2] first commit
    1 file changed, 1 insertion(+), 1 deletion(-)
    $ copier . ../copier-troubleshoot-task-gen
    No git tags found in template; using HEAD as ref
    Either c15ca61 or None is not a PEP 440 valid version.
    
     conflict  copier_post_generation_task.py
     # Press <return> for Yes to overwrite
     Overwrite /Users/emmanuelsciara/Documents/Development/git-repos/veolia/copier-troubleshoot-task-gen/copier_post_generation_task.py? [Y/n]
        force  copier_post_generation_task.py
     conflict  .copier-answers.yml
     # Press <return> for Yes to overwrite
     Overwrite /Users/emmanuelsciara/Documents/Development/git-repos/veolia/copier-troubleshoot-task-gen/.copier-answers.yml? [Y/n]
        force  .copier-answers.yml
    
    > Running task 1 of 1: python3 copier_post_generation_task.py
    You really ran 'copier_post_generation_task.py'!
    
    ```
    </details>
6. commit the newly generated files:
    <details>

    ```bash
    $ pushd ../copier-troubleshoot-task-gen
    $ git add .
    $ git commit -m'second template generation'
    [main d7c6558] to be squashed
    2 files changed, 5 insertions(+), 1 deletion(-)
    $ popd
    ```
    </details>
7. Change the `copier.yaml` file to add a prompt:
    ```yaml
    # CONFIGURATION -------------------------
    _subdirectory: project
    _tasks:
      - python3 copier_post_generation_task.py
    
    prompt:
    type: str
    help: Test prompt
    default: default
    ```
6. Commit the changes and re-generate the project:
    <details>

    ```bash
    # In the `copier-troubleshoot-task/` directory
    $ git add .
    $ git commit -m'copier.yaml with prompt'
    [main e97c185] to be squashed
    1 file changed, 5 insertions(+)
    $ copier . ../copier-troubleshoot-task-gen
    No git tags found in template; using HEAD as ref
    
    Test prompt
    prompt? Format: str
    🎤 [default]:
    Either 8605f79 or None is not a PEP 440 valid version.
    You really ran 'copier_post_generation_task.py'!
    
     identical  copier_post_generation_task.py
      conflict  .copier-answers.yml
     Overwrite /Users/emmanuelsciara/Documents/Development/git-repos/veolia/copier-troubleshoot-task-gen/.copier-answers.yml? [Y/n]
         force  .copier-answers.yml
    
    > Running task 1 of 1: python3 copier_post_generation_task.py
    You really ran 'copier_post_generation_task.py'!
    
   ```
    </details>

During the last project generation, the task was run before and after the generation of the files.

More over, if the task script file was changed in the template, it is the old version of the task script file that is run before the files generation and the new version that is run after it.

**Logs**

See steps to reproduce.

**Expected behavior**

Task commands should be executed once, in their latest version, after generating or updating a project from the template.

**Screenshots/screencasts/logs**

N/A

**Environment**
- OS: macOS (M1) 12.0.1
- Copier version: `5.1.0`
- Python version: `3.10.0`
- Installation method: `pipx+pypi` (`pipx` was installed with `homebrew`, hence with version `python@3.10`)

**Additional context**

N/A

**Comment by Yajo** [comment]
In #240 there's a discussion about how tasks should work. It should be refactored because current workflow has these corner cases that show its weaknesses. 

I'll close as duplicate of that one, as the refactor will fix this problem when it lands.
